pkg/ruler/rule_backup_manager.go

```diff
@@ -0,0 +1,150 @@
+package ruler
+
+import (
+	"context"
+	"errors"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/prometheus/model/rulefmt"
+	"github.com/prometheus/prometheus/promql/parser"
+	promRules "github.com/prometheus/prometheus/rules"
+
+	"github.com/cortexproject/cortex/pkg/ruler/rulespb"
+)
+
+// Implements GroupLoader interface but instead of reading from a file when Load is called, it returns the
+// rulefmt.RuleGroup it has stored
+type loader struct {
+	ruleGroups map[string][]rulefmt.RuleGroup
+}
+
+func (r *loader) Load(identifier string) (*rulefmt.RuleGroups, []error) {
+	return &rulefmt.RuleGroups{
+		Groups: r.ruleGroups[identifier],
+	}, nil
+}
+
+func (r *loader) Parse(query string) (parser.Expr, error) {
+	return parser.ParseExpr(query)
+}
+
+// rulesBackupManager is an in-memory store that holds []promRules.Group of multiple users. It only stores the Groups,
+// it doesn't evaluate them.
+type rulesBackupManager struct {
+	inMemoryRuleGroupsBackup map[string][]*promRules.Group
+	cfg                      Config
+
+	logger log.Logger
+
+	backupGroupRulesCount      *prometheus.GaugeVec
+	lastBackupReloadSuccessful *prometheus.GaugeVec
+}
+
+func newRulesBackupManager(cfg Config, logger log.Logger, reg prometheus.Registerer) *rulesBackupManager {
+	return &rulesBackupManager{
+		inMemoryRuleGroupsBackup: make(map[string][]*promRules.Group),
+		cfg:                      cfg,
+		logger:                   logger,
+
+		backupGroupRulesCount: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: "cortex",
+			Name:      "ruler_backup_rule_group_rules",
+			Help:      "The number of backed up rules",
+		}, []string{"user", "rule_group"}),
+		lastBackupReloadSuccessful: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: "cortex",
+			Name:      "ruler_backup_last_reload_successful",
+			Help:      "Boolean set to 1 whenever the last configuration reload attempt was successful.",
+		}, []string{"user"}),
+	}
+}
+
+// setRuleGroups updates the map[string][]*promRules.Group that the rulesBackupManager stores in memory.
+func (r *rulesBackupManager) setRuleGroups(_ context.Context, ruleGroups map[string]rulespb.RuleGroupList) {
+	backupRuleGroups := make(map[string][]*promRules.Group)
+	for user, groups := range ruleGroups {
+		promGroups, err := r.ruleGroupListToPromGroups(user, groups)
```

Comment:
I am trying to understand why we need to convert `rulespb.RuleGroupList` to prometheus rules type and store them in memory. Why we cannot store `rulespb.RuleGroupList` directly?

At the end of the day, when we call `getShardedRules` and `getLocalRules`, the final response is `rulespb` type defined by us.
If we store Prometheus type here we are basically converting twice, both at read and write time.

And I think `promManager.LoadGroups` is very expensive since it needs to run `m.opts.RuleDependencyController.AnalyseRules(rules)` to do the topology sort every time.

Reply:
I believe we have to convert the `rulespb.RuleGroupList` to `promRules.Group` because

1. I am hoping to reuse the prometheus code that sets the default value of fields like group`s `Interval`,  field. We could make the code convert  `rulespb.RuleGroupList` to `GroupStateDesc` which is what `getLocalRules` and just make sure to copy behavior from https://github.com/prometheus/prometheus/blob/main/rules/manager.go#L280-L343
2. By converting it to groups and appending it to the list returned by the manager, we can reuse the existing code for supporting filters https://github.com/cortexproject/cortex/blob/master/pkg/ruler/ruler.go#L778-L798 . We could write a different code path that filters `GroupStateDesc`

I did not realize that `m.opts.RuleDependencyController.AnalyseRules(rules)` can be expensive. Since this is only relevant for evaluation we could get around it by passing in a noop `RuleDependencyController`.

I think converting to prometheus rules type is the simpler approach. But I can change it if you think otherwise.

Reply:
> I did not realize that m.opts.RuleDependencyController.AnalyseRules(rules) can be expensive. Since this is only relevant for evaluation we could get around it by passing in a noop RuleDependencyController.

I am good if we pass a noop controller.

Reply:
> By converting it to groups and appending it to the list returned by the manager, we can reuse the existing code for supporting filters https://github.com/cortexproject/cortex/blob/master/pkg/ruler/ruler.go#L778-L798

I don't see how this code is for prometheus rules only. It looks very general, no?

> I am hoping to reuse the prometheus code that sets the default value of fields like groups Interval,  field. We could make the code convert  rulespb.RuleGroupListtoGroupStateDescwhich is whatgetLocalRules` return and just make sure to copy behavior from https://github.com/prometheus/prometheus/blob/main/rules/manager.go#L280-L343

The reason why I don't want to reuse this code is because I don't want to initialize a new rules manager everytime when this code path is called. It is expensive since it initializes rules metrics everytime but not using them at all.
Since what we need is to just use the rules we cached in memory, can we just use the loader directly without initializing the manager?

The way I can think of is to copy paste the code out and remove manager dependency. However, if we do this, why we don't just store rulespb.RuleGroupList?

Reply:
> The reason why I don't want to reuse this code is because I don't want to initialize a new rules manager everytime when this code path is called. It is expensive since it initializes rules metrics everytime but not using them at all.

I am ok if we pre-initialize the manager metrics and pass it every time.

Reply:
I see, I agree. It is more optimal to store `rulespb.RuleGroupList` and do the conversion to `GroupStateDesc` upon request. I will make the change. Thank you :bow: